Support for importers plugins

aperezhortal opened the pull request:
This PR adds support for adding additional importers to the pysteps interface without modifying the pysteps source code.  Instead, the users can add custom importers to pysteps by installing external packages, called plugins, that are automatically "discovered" and integrated to pysteps.

How does it work?
--------------------------

Let's suppose that we have a plugin (python package), named `importer_abc_xyz`, that contains the following importers:
`importer_abc_xyz.importers.import_abc_yyy`
`importer_abc_xyz.importers.import_abc_zzz`

When the plugin is installed, it adds [entry_points](https://packaging.python.org/specifications/entry-points/) to the python installation advertising the components of this plugin to other packages (in our case, pysteps). 

Next, when pysteps is imported, it [discovers the available entry_points](https://packaging.python.org/guides/creating-and-discovering-plugins/#using-package-metadata) that belong to the **pysteps.plugins.importers** group (the group name is defined in the plugin). These discovered importers are added as attributes to the `io.importers` module, and they are registered to the `io.get_method` interface. 

After the external importers are registered, they become part of the pysteps interface and can be used together with the other available importers. 

Plugin project template
------------------------------

To make the creation of the plugin easier, a [cookiecutter](https://cookiecutter.readthedocs.io/en/1.7.2/README.html) project template, available at https://github.com/pySTEPS/cookiecutter-pysteps-plugin, can be used to quickly create a plugin project by executing:
`$> cookiecutter https://github.com/pySTEPS/cookiecutter-pysteps-plugin`
and follow the interactive instructions.

Documentation
-----------------

- Plugin template documentation: https://cookiecutter-pysteps-plugin.readthedocs.io/en/latest/
- Pysteps documentation (plugins section): https://pysteps.readthedocs.io/en/importer_plugins_support/developer_guide/importer_plugins.html


PR progress [Updated]
---------------------------------
- [x] Add plugin support
- [x] Minimal working example
- [x] Add developers guidelines to develop custom plugins [(docs here)](https://pysteps.readthedocs.io/en/importer_plugins_support/developer_guide/importers_plugins.html)
- [x] Move the plugin cookiecutter-template to the PySTEPS github
- [x] Improve the plugin cookiecutter-template documentation
- [x] Add more tests (multiple plugins, tests unexpected behaviors, etc.)

aperezhortal committed:
Add importers plugins functionality

aperezhortal committed:
Add supports for plugins adding new importers

aperezhortal committed:
Test the plugins in Travis

The plugin functionality it is only tested in the minimal checks.

aperezhortal committed:
Ignore IDE files and clean unused entries

dnerini commented:
Hi @aperezhortal 
This looks like a great idea!
I still have to familiarize myself with the details and give it a try, but I like it already ;-)

So, if I understand correctly, to install a new custom importer, I would have to first run the cookiecutter, then implement my importer in the template, right? After this, how is the registration into pysteps done? Simply using pip install?

One more question, I assume new plugins won't be shared with the rest of the community, correct? These will essentially remain private methods.

Once again, great idea, thanks!

aperezhortal commented:
> So, if I understand correctly, to install a new custom importer, I would have to first run the cookiecutter, then implement my importer in the template, right? 

The first step is creating the plugin (a python package). To facilitate this, cookiecutter (and the cookiecutter-pysteps-plugin template) is used to create the skeleton of the plugin project. Then, the new importers need to be added to the barebone project.

> After this, how is the registration into pysteps done? Simply using pip install?

When you install the package using `pip install` or `python setup.py install`, it add the entry points defined in the setup.py script to the python distribution. The installation adds the entry points a .txt file located in, for example,  **<path_to_my_python_environment>/site-packages/<name_of_my_importer_plugin>-<version>.dist-info/entry_points.txt**. 

Then, when pysteps is importer, it calls the pysteps.io.interface.discover_importers() function that looks for the entry points available in the python distribution and registers the ones that belong to the "pysteps.plugins.importers" group (the group name was defined in the setup.py).

The main advantage of this approach is that the plugins will keep working if the pysteps version is updated without any user intervention.

> One more question, I assume new plugins won't be shared with the rest of the community, correct? These will essentially remain private methods.

This is the best part, in my opinion. The users who create the plugins have ownership and they are free to distribute or not the packages. For example, if an organization uses "in-house" tools to read files that cannot be openly distributed, they can internally distribute the plugin.

*(I will include the information of these anwers in the plugins documentation)*

dnerini commented:
This sounds really great, I personally think we should go ahead. What would be the next steps? How can we help?

aperezhortal commented:
Great! 
The plugin support is already in place. The next step should be writing a tutorial to add a new plugin and write the template documentation. Help with this is more than welcome :bowtie: 

I will push the first draft soon with the minimal installation instructions to use it as a starting point. We can then incrementally improve the tutorial to make it as clear and easy to follow as possible.

aperezhortal committed:
Improve decorator docstrings

aperezhortal committed:
Add decorators to reference docs

aperezhortal committed:
Correct the function name in the autosummary

aperezhortal committed:
Fix indentation in postprocess_import docs

aperezhortal committed:
Add importers plugins documentation

This is the first draft for the importers plugin implementation.

aperezhortal committed:
Add cookiecutter template instructions

aperezhortal committed:
Fix typo

dnerini commented on doc/source/developer_guide/importers_plugins.rst:
```diff
@@ -0,0 +1,282 @@
+Create your own importers plugin
+================================
+
+Since version 3.4, pysteps allows the users to add new importers by installing external
```
```suggestion
Since version 1.4, pysteps allows the users to add new importers by installing external
```
I suppose this feature will be included in release 1.4.

aperezhortal replied:
Oops, I meant 1.4 :laughing:

dnerini commented on doc/source/developer_guide/importers_plugins.rst:
```diff
@@ -0,0 +1,282 @@
+
+When the plugin is installed, it advertises the new importers to other packages (in our
+case, pysteps) using the python `entry points specification`_.
+This new importers are automatically discovered every time that the pysteps library is
```
```suggestion
These new importers are automatically discovered every time that the pysteps library is
```

dnerini commented on doc/source/developer_guide/importers_plugins.rst:
```diff
@@ -0,0 +1,282 @@
+One is building the importers plugin package from scratch.
+An easier alternative is using a `Cookiecutter`_ template that easily create the skeleton
+for an importer plugin template.
+Although the latter alternative is recommended (the cookiecutter template),
```
```suggestion
Although the cookiecutter template is recommended,
```

dnerini commented on doc/source/developer_guide/importers_plugins.rst:
```diff
@@ -0,0 +1,282 @@
+        ├── pysteps_importer_abc
+        └───── __init__.py
+
+The __init__.py files are required inform python that a given directory contains a
```
```suggestion
The __init__.py files are required to inform python that a given directory contains a
```

dnerini commented on doc/source/developer_guide/importers_plugins.rst:
```diff
@@ -0,0 +1,282 @@
+        └───── importer_abc_xyz.py  (importer module)
+
+Inside the package folder (*pysteps_importer_abc*), we place the python module
+(or modules) containing the importers' implementation.
```
```suggestion
(or modules) containing the actual implementation of our new importers.
```

dnerini commented on doc/source/developer_guide/importers_plugins.rst:
```diff
@@ -0,0 +1,282 @@
+https://github.com/aperezhortal/cookiecutter-pysteps-plugin
+.
+
+The first step is two install the latest version of `Cookiecutter`_ using,
```
```suggestion
The first step is to install the latest version of `Cookiecutter`_ using,
```

dnerini commented on doc/source/developer_guide/importers_module_example.py:
```diff
@@ -0,0 +1,251 @@
+    ####################################################################################
+    # The {extra_kwargs_doc} above is needed to add default keywords added to this     #
+    # importer by the pysteps.decorator.postprocess_import decorator.                  #
+    # IMPORTANT: Remove these box in the final version of this function                #
```
```suggestion
    # IMPORTANT: Remove this box in the final version of this function                 #
```

dnerini commented on doc/source/developer_guide/importers_module_example.py:
```diff
@@ -0,0 +1,251 @@
+#     PYPROJ_IMPORTED = False
+
+from pysteps.decorators import postprocess_import
+
```
```suggestion
from pysteps.exceptions import MissingOptionalDependency

```

Add this in case optional dependencies are needed? Maybe we could include as commented line?

dnerini commented on doc/source/developer_guide/importers_module_example.py:
```diff
@@ -0,0 +1,251 @@
+# convention:
+# https://pysteps.readthedocs.io/en/latest/pysteps_reference/io.html#available-importers
+#
+# The function prototype for the importer's declaration should have the following form:
```
What do you mean for function prototype? Can we simplify this sentence with something like "The definition of your importer function should have the following form" ?

aperezhortal replied:
By prototype, I was referring to the function's form, but it was indeed redundant and confusing. I changed the sentence as you suggested (commit 61052e2).

dnerini commented on doc/source/developer_guide/importers_module_example.py:
```diff
@@ -0,0 +1,251 @@
+    #     )
+
+    ####################################################################################
+    # Add the code to read the precipitation data here. Note that only cartesian grid
```
```suggestion
    # Add the code to read the precipitation data here. Note that only cartesian grids
```

dnerini commented on doc/source/developer_guide/importers_module_example.py:
```diff
@@ -0,0 +1,251 @@
+    # grib file stores the data using double precision.
+
+    # Quality field, should have the same dimensions of the precipitation field.
+    # Use None is not information is available.
```
```suggestion
    # Use None if not information is available.
```

dnerini commented on doc/source/developer_guide/importers_module_example.py:
```diff
@@ -0,0 +1,251 @@
+    # For additional information on the metadata fields, see:
+    # https://pysteps.readthedocs.io/en/latest/pysteps_reference/io.html#pysteps-io-importers
+
+    # The projection definition is an string with a PROJ.4-compatible projection
```
```suggestion
    # The projection definition is a string with a PROJ.4-compatible projection
```

dnerini commented on doc/source/developer_guide/importers_module_example.py:
```diff
@@ -0,0 +1,251 @@
+    return precip, quality, metadata
+
+
+# Function import_abc_netcdf to import NETCDF files from the Awesome Bureau of Composites
```
Why you need to include it another time?

dnerini replied:
got it, it's the netcdf version... duh...

dnerini commented on doc/source/developer_guide/importers_plugins.rst:
```diff
@@ -0,0 +1,282 @@
+.. _importers-plugins:
```
```suggestion
.. _importer-plugins:
```
A detail, but I'd suggest to rename it with a singular importer...

dnerini commented on doc/source/developer_guide/importers_plugins.rst:
```diff
@@ -0,0 +1,282 @@
+======================
+
+There are two ways of creating your own plugin.
+One is building the importers plugin package from scratch.
```
```suggestion
The first one involves building the importers plugin package from scratch.
```

dnerini commented on doc/source/developer_guide/importers_plugins.rst:
```diff
@@ -0,0 +1,282 @@
+There are two ways of creating your own plugin.
+One is building the importers plugin package from scratch.
+An easier alternative is using a `Cookiecutter`_ template that easily create the skeleton
+for an importer plugin template.
```
```suggestion
for the new importer plugin.
```

dnerini commented on doc/source/developer_guide/importers_plugins.rst:
```diff
@@ -0,0 +1,282 @@
+Let's suppose that we want to add two new importers to pysteps for reading the radar
+composites from the "Awesome Bureau of Composites", kindly abbreviated as "abc".
+The composites provided by this institution are available in two different
+formats: Netcdf and Grib2. The details of each format is not important for the rest of
```
```suggestion
formats: Netcdf and Grib2. The details of each format are not important for the rest of
```

dnerini commented on doc/source/developer_guide/setup.py:
```diff
@@ -0,0 +1,91 @@
+# projects. See more at: https://setuptools.readthedocs.io/en/latest/
+from setuptools import setup, find_packages
+
+# The long description is used to explain to the users how the package is install and
```
```suggestion
# The long description is used to explain to the users how the package is installed and
```

dnerini committed:
Apply suggestions from code review

@aperezhortal this is truly amazing work! The documentation is excellent, I just corrected few typos and suggested few changes.

dnerini committed:
Update and rename importers_plugins.rst to importer_plugins.rst

dnerini committed:
Update index.rst

aperezhortal committed:
Minor corrections

aperezhortal committed:
Improve description on importer's function form

aperezhortal committed:
Move template to pySTEPS github

aperezhortal committed:
Update tests for plugins

aperezhortal committed:
Fix typo

aperezhortal committed:
Merge with main branch

aperezhortal committed:
Update plugins' documentation

aperezhortal committed:
Remove non-existing function from autosummary

aperezhortal committed:
Define the bib file in bibtex_bibfiles

aperezhortal committed:
Fix sphinx warnings

aperezhortal committed:
Add plugins tests

aperezhortal committed:
Add cookiecutter test dependency

dnerini reviewed (approved):
Very nice, Andres, very cool contribution!

aperezhortal committed:
Remove debug messages

aperezhortal committed:
Add :all: option to bibliography

RubenImhoff reviewed (approved):
Great work @aperezhortal! I have nothing to add. :)
